src/librustc_typeck/mem_categorization.rs

```diff
@@ -679,20 +762,67 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         op(&place_with_id, pat);
 
         match pat.kind {
-            PatKind::TupleStruct(_, ref subpats, _) | PatKind::Tuple(ref subpats, _) => {
-                // S(p1, ..., pN) or (p1, ..., pN)
+            PatKind::Tuple(ref subpats, _) => {
+                // (p1, ..., pN)
                 for subpat in subpats.iter() {
                     let subpat_ty = self.pat_ty_adjusted(&subpat)?;
-                    let sub_place = self.cat_projection(pat, place_with_id.clone(), subpat_ty);
+                    let sub_place = self.cat_projection(
+                        pat,
+                        place_with_id.clone(),
+                        subpat_ty,
+                        ProjectionKind::Other,
+                    );
+                    self.cat_pattern_(sub_place, &subpat, op)?;
+                }
+            }
+
+            PatKind::TupleStruct(ref qpath, ref subpats, dots_pos) => {
+                // S(p1, ..., pN)
+                let variant_index = self.variant_index_for_adt(qpath, pat.hir_id, pat.span)?;
+                let total_fields =
+                    self.total_fields_in_adt_variant(pat.hir_id, variant_index, pat.span)?;
+
+                for (i, subpat) in subpats.iter().enumerate() {
+                    let subpat_ty = self.pat_ty_adjusted(&subpat)?;
+                    let projection_kind = if let Some(dots_pos) = dots_pos {
+                        if i < dots_pos {
```

Comment:
Ugh, I don't know a better option, but it feels like a shame to reproduce this logic. I'm sure there is code like this sprinkled all throughout the compiler...

Reply:
https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir/pat_util.rs.html#33

This exists